components/Questionnaire/Page.js

```diff
@@ -0,0 +1,376 @@
+                  {userAnswerCount < 1
+                    ? t('questionnaire/invalid')
+                    : <A href='#' onClick={this.handleReset}>{t('questionnaire/cancel')}</A>
+                  }
+                </div>
+              </div>
+            </div>
+          )
+        }} />
+      </Frame>
+    )
+  }
+}
+
+const submitAnswerMutation = gql`
+mutation submitAnswer($questionId: ID!, $payload: JSON) {
+  submitAnswer(answer: {
+    questionId: $questionId,
+    payload: $payload
+  }) {
+    ... on QuestionInterface {
+      id
+      userAnswer {
+        id
+        payload
+      }
+    }
+  }
+}
+`
+
+const resetQuestionnaireMutation = gql`
+mutation resetQuestionnaire($id: ID!) {
+  resetQuestionnaire(id: $id) {
+    id
+  }
+}
+`
+
+const submitQuestionnaireMutation = gql`
+mutation submitQuestionnaire($id: ID!) {
+  submitQuestionnaire(id: $id) {
+    id
+    userSubmitDate
+    userHasSubmitted
+  }
+}
+`
+
+const query = gql`
+query getQuestionnaire($slug: String!) {
+  questionnaire(slug: $slug) {
+    id
+    beginDate
+    endDate
+    userHasSubmitted
+    userSubmitDate
+    questions {
+      ... on QuestionInterface {
+        id
+        order
+        text
+        userAnswer {
+          id
+          payload
+        }
+      }
+      ... on QuestionTypeText {
+        maxLength
+      }
+      ... on QuestionTypeChoice {
+        cardinality
+        options {
+          label
+          value
+          category
+        }
+      }
+      ... on QuestionTypeRange {
+        kind
+        ticks {
+          label
+          value
+        }
+      }
+      ... on QuestionTypeDocument {
+        template
+      }
+    }
+  }
+}
+`
+
+export default compose(
+  withT,
+  withRouter,
+  graphql(submitQuestionnaireMutation, {
+    props: ({ mutate }) => ({
+      submitQuestionnaire: (id) => {
+        return mutate({
+          variables: {
+            id
+          }
+        })
+      }
+    })
+  }),
+  graphql(resetQuestionnaireMutation, {
+    props: ({ mutate, ownProps: { router } }) => ({
+      resetQuestionnaire: (id) => {
+        return mutate({
+          variables: {
+            id
+          },
+          refetchQueries: [{ query, variables: { slug: router.query.slug } }]
+        })
+      }
+    })
+  }),
+  graphql(submitAnswerMutation, {
+    props: ({ mutate }) => ({
+      submitAnswer: (questionId, payload, answerId) => {
+        const optimistic = {
+          optimisticResponse: {
+            __typename: 'Mutation',
+            submitAnswer: {
+              __typename: 'QuestionInterface',
+              id: questionId,
+              userAnswer: {
+                __typename: 'Answer',
+                id: answerId,
+                payload
+              }
+            }
+          }
+        }
+        return mutate({
+          variables: {
+            questionId,
+            payload
+          },
+          ...(answerId ? optimistic : {})
```

Comment:
this leads to initial race conditions. we should do client side uuid generation, let's discuss with @patte